### ISSUE TITLE
Is there a fault in heap_sort?

### ISSUE THREAD

**Issue by Jiangluyu** [matthewsamuel95/ACM-ICPC-Algorithms#1036]
I have prudently went through the source code, but I found a problem(maybe) when I use the following data: {42, 20, 17, 13, 28, 14, 23, 15}, I think there are other WRONG data like this. In my opinion, that's because in the last sort loop, the arr is 28 42 23 20 17 15 14 13, and the heap left is 28 42, so it does not satisfy the condition to swap 28 & 42, but actually 28 42 is not in the right order.

**Comment by anishsingh935** [comment]
can I fix this issues please

**Comment by yamanalrfai** [comment]
Yeah, You right in a C++ version of heap_sort there are 3 wrongs  

1. the function minheapify (line: 49)
   `if (l >= limit || r >= limit)`
   `return;`
   using `||` is not right because if the `r >= limit` then it never check the left node.
   so we must cheek like this:

```cpp
void minheapify(int i, int limit) {
  int l = left(i);
  int r = right(i);
  int smallest = i;
  if (l < limit && harr[l] < harr[i])
    smallest = l;
  if (r < limit && harr[r] < harr[smallest])
    smallest = r;
  if (i != smallest) {
    swap( & harr[i], & harr[smallest]);
    minheapify(smallest, limit);
  }
}
```

2. the function sort() (line: 64)
   `for (i = 1; i < heap_size - 1; i++) {`
   it must be 
   `for (i = 1; i < heap_size; i++) {`

3. need to add buildHeap before start the sort so it become 

```cpp
void buildHeap() {
  for (int i = (heap_size / 2) - 1; i >= 0; i--) {
    minheapify(i, heap_size);
  }
}

void sort() {
  buildHeap(); 
  int i;
  for (i = 1; i < heap_size; i++) {
    swap( & harr[0], & harr[heap_size - i]);
    minheapify(0, heap_size - i);
  }
}
```

and for Memory leek use Destructor

```cpp
~heap() {
  delete[] harr;
}
```

I will not add this Because your class manages a raw memory pointer (harr), if you ever tried to copy this heap like this:
`heap h2 = h;`
...both heaps would point to the exact same array in memory. Modifying one would break the other, and when they are destroyed, the program would try to delete[] the same memory twice, causing a crash.

I will open a PR to add this thing 
here it if still you need it after 6 years : https://github.com/matthewsamuel95/ACM-ICPC-Algorithms/pull/1060/changes
